UnityAds/UnityFullscreenAdAdapter.m

```diff
@@ -0,0 +1,184 @@
+        [self.delegate fullscreenAdAdapter:self didFailToLoadAdWithError:error];
+        return;
+    }
+
+    // Only need to cache game ID for SDK initialization
+    [UnityAdsAdapterConfiguration updateInitializationParameters:info];
+
+    NSString *placementId = [info objectForKey:kUnityAdsOptionPlacementIdKey];
+    if (placementId == nil) {
+        placementId = [info objectForKey:kUnityAdsOptionZoneIdKey];
+    }
+
+    if (placementId == nil) {
+        NSError *error = [NSError errorWithAdAdapterErrorCode: MPAdAdapterErrorCodeInvalidAdapter
+                                                  description: @"Custom event class data did not contain placementId."
+                                                   suggestion: @"Update your MoPub custom event class data to contain a valid Unity Ads placementId."];
+        MPLogAdEvent([MPLogEvent adLoadFailedForAdapter:NSStringFromClass(self.class) error:error], placementId);
+        [self.delegate fullscreenAdAdapter:self didFailToLoadAdWithError:error];
+        return;
+    }
+    
+    UADSLoadOptions *options = [UADSLoadOptions new];
+    if (adMarkup != nil) {
+        _objectId = [[NSUUID UUID] UUIDString];
+        [options setObjectId:_objectId];
+        [options setAdMarkup:adMarkup];
+    }
+    
+    MPLogAdEvent([MPLogEvent adLoadAttemptForAdapter:NSStringFromClass(self.class) dspCreativeId:nil dspName:nil], placementId);
+    [UnityAds load:placementId options:options loadDelegate:self];
+}
+
+- (void)presentAdFromViewController:(UIViewController *)viewController
+{
+    if (![self hasAdAvailable]) {
+        MPLogWarn(@"Unity Ads received call to show before successfully loading an ad");
+    }
+    
+    UADSShowOptions *options = [UADSShowOptions new];
+    if (_objectId != nil) {
+        [options setObjectId:_objectId];
+    }
+   
+    MPLogAdEvent([MPLogEvent adShowAttemptForAdapter:NSStringFromClass(self.class)], [self getAdNetworkId]);
+    [UnityAds show:viewController placementId:_placementId options:options showDelegate:self];
+}
+
+- (void)handleDidInvalidateAd
+{
+  // Nothing to clean up.
+}
+
+#pragma mark - UnityAdsLoadDelegate Methods
+
+- (void)unityAdsAdLoaded:(nonnull NSString *)placementId {
+    self.placementId = placementId;
+    MPLogAdEvent([MPLogEvent adLoadSuccessForAdapter:NSStringFromClass(self.class)], placementId);
+    [self.delegate fullscreenAdAdapterDidLoadAd:self];
+}
+
+- (void)unityAdsAdFailedToLoad:(nonnull NSString *)placementId withError:(UnityAdsLoadError)error withMessage:(NSString *)message {
+    self.placementId = placementId;
+
+    NSError *loadError = [NSError errorWithAdAdapterErrorCode:MPAdAdapterErrorCodeUnknown
+                                                  description:[NSString stringWithFormat:@"%@ ad failed to load a video ad for %@, with error message: %@", self.adType, placementId, message]];
```

Comment:
nit: "ad failed to load a video ad" -> "failed to load a __adtype__ video ad for placement __placementId___"